libcontainer/nsenter/cloned_binary.c

```diff
@@ -627,14 +472,6 @@ static int clone_binary(void)
 	size_t sent = 0;
 	int fdtype = EFD_NONE;
 
-	/*
-	 * Before we resort to copying, let's try creating an ro-binfd in one shot
-	 * by getting a handle for a read-only bind-mount of the execfd.
-	 */
-	execfd = try_bindfd();
```

Comment:
As in do something like `#ifdef WANT_NSEXEC_BINDFD`? We could, and maybe we should do the deprecation that way (keep it enabled but emit a warning for `1.1.x`, disable it for `1.2.0`, and then remove it for `1.3.0`?).

As for the memory usage, I want to reiterate that the memory usage is _temporary_ -- for most users, `runc init` is only alive for a few hundred milliseconds and then dies soon after. AFAIK `mmap`'d file-backed memory is not accounted by memcg, so there would be no double-accounting -- the total extra memory usage would be 10MB per `runc init` process (for the 100-300ms that the process is alive). Based on my tests, setting the limit to >13MB (the current size of the `runc` binary) is enough for containers to be able to start. However, if you were to do 5000 `runc exec`s concurrently, there could be significant memory usage for the short period of time those processes are running.

There are a few workarounds for this, but they aren't particularly pretty. The simplest one is to create a bind-mount for a sealed memfd for runc to use, however (because you cannot bind-mount the memfd directly) you have to have a long-running process that keeps the magic-link to a memfd alive. This way there will be no additional memory usage at all (because the memfd was created in the host's memcg). Here is a program that implements this feature (I might add this to `contrib` actually...):

```go
/*
 * Copyright (c) 2023 SUSE LLC
 * Copyright (c) 2023 Aleksa Sarai <cyphar@cyphar.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package main

import (
	"errors"
	"fmt"
	"io"
	"os"
	"strings"
	"time"

	"github.com/urfave/cli"
	"golang.org/x/sys/unix"
)

// version will be populated by the Makefile, read from
// VERSION file of the source code.
var version = ""

// gitCommit will be the hash that the binary was built from
// and will be populated by the Makefile
var gitCommit = ""

const (
	usage = `Open Container Initiative contrib/cmd/memfdexe

In order to protect against certain container attacks, every runc invocation
that involves creating or joining a container will cause runc to make a copy of
the runc binary in memory (usually to a memfd). While "runc init" is very
short-lived, this extra memory usage can cause problems for containers with
very small memory limits (or containers that have many "runc exec" invocations
applied to them at the same time).

memfdexe is a tool to create a persistent memfd-sealed-copy of the runc binary,
which will cause runc to not make its own copy. This means you can get the
benefits of using a sealed memfd as runc's binary (even in a container breakout
attack to get write access to the runc binary, neither the underlying binary
nor the memfd copy can be changed).

To use memfdexe, just specify which path you want to create a socket path at
which you want to receive terminals:

    $ memfdexe /usr/bin/runc

Note that (due to kernel restrictions on bind-mounts), this program must remain
running on the host in order for the binary to be readable (it is recommended
you use a systemd unit to keep this process around).

If this program dies, there will be a leftover mountpoint that always returns
-EINVAL when attempting to access it. You need to use memfdexe --cleanup on the
path in order to unmount the path (regular umount(8) will not work):

    $ memfdexe --cleanup /usr/bin/runc
`
)

func cleanup(path string) error {
	fd, err := unix.Open(path, unix.O_PATH|unix.O_NOFOLLOW|unix.O_CLOEXEC, 0)
	if err != nil {
		return fmt.Errorf("cleanup: failed to open runc binary path: %w", err)
	}
	fdPath := fmt.Sprintf("/proc/self/fd/%d", fd)

	// Keep umounting until we hit a umount error.
	for unix.Unmount(fdPath, unix.MNT_DETACH) == nil {
		// loop...
	}
	return nil
}

func memfdClone(path string) (*os.File, error) {
	binFile, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("memfd clone: failed to open runc binary path: %w", err)
	}
	defer binFile.Close()

	const memfdComment = "runc_cloned:/proc/self/exe"
	memfd, err := unix.MemfdCreate(memfdComment, unix.MFD_CLOEXEC|unix.MFD_EXEC|unix.MFD_ALLOW_SEALING)
	if errors.Is(err, unix.EINVAL) {
		// old kernel with no MFD_EXEC support
		memfd, err = unix.MemfdCreate(memfdComment, unix.MFD_CLOEXEC|unix.MFD_ALLOW_SEALING)
	}
	if err != nil {
		return nil, fmt.Errorf("memfd clone: failed to create memfd: %w", err)
	}
	memfdFile := os.NewFile(uintptr(memfd), memfdComment)
	// we *do not* defer memfdFile.Close() here...

	// Copy the contents...
	for {
		_, err = io.Copy(memfdFile, binFile)
		if !errors.Is(err, unix.EINTR) {
			break
		}
	}
	if err != nil {
		return nil, fmt.Errorf("memfd clone: failed to copy binary text to memfd: %w", err)
	}

	// Seal the memfd. We ignore errors for the newer seals, but for the
	// RUNC_MEMFD_MIN_SEALS set we will return an error. This matches the
	// behaviour of libcontainer/nsenter/cloned_binary.c.
	const RUNC_MEMFD_MIN_SEALS = unix.F_SEAL_SEAL | unix.F_SEAL_SHRINK | unix.F_SEAL_GROW | unix.F_SEAL_WRITE
	const F_SEAL_EXEC = 0x20 // from <linux/fcntl.h>
	_, _ = unix.FcntlInt(uintptr(memfd), unix.F_ADD_SEALS, unix.F_SEAL_FUTURE_WRITE)
	_, _ = unix.FcntlInt(uintptr(memfd), unix.F_ADD_SEALS, F_SEAL_EXEC)
	_, err = unix.FcntlInt(uintptr(memfd), unix.F_ADD_SEALS, RUNC_MEMFD_MIN_SEALS)
	if err != nil {
		return nil, fmt.Errorf("memfd clone: failed to add required seals to memfd: %w", err)
	}
	return memfdFile, nil
}

func mount(path string) error {
	memfdFile, err := memfdClone(path)
	if err != nil {
		return err
	}
	defer memfdFile.Close()
	memfdPath := fmt.Sprintf("/proc/self/fd/%d", memfdFile.Fd())

	// We have to open an O_NOFOLLOW|O_PATH to the memfd magic-link because we
	// cannot bind-mount the memfd itself (it's in the internel kernel mount
	// namespace and cross-mount-namespace bind-mounts are not allowed). This
	// also requires that this program stay alive continuously for the
	// magic-link to stay alive...
	memfdLinkFd, err := unix.Open(memfdPath, unix.O_PATH|unix.O_NOFOLLOW|unix.O_CLOEXEC, 0)
	if err != nil {
		return fmt.Errorf("mount: failed to /proc/self/fd magic-link for memfd: %w", err)
	}
	defer unix.Close(memfdLinkFd)
	memfdLinkFdPath := fmt.Sprintf("/proc/self/fd/%d", memfdLinkFd)

	exeFd, err := unix.Open(path, unix.O_PATH|unix.O_NOFOLLOW|unix.O_CLOEXEC, 0)
	if err != nil {
		return fmt.Errorf("mount: failed to open target runc binary path: %w", err)
	}
	exeFdPath := fmt.Sprintf("/proc/self/fd/%d", exeFd)
	defer unix.Close(exeFd)

	err = unix.Mount(memfdLinkFdPath, exeFdPath, "", unix.MS_BIND, "")
	if err != nil {
		return fmt.Errorf("mount: failed to mount memfd on top of runc binary path target: %w", err)
	}
	defer cleanup(path) // should never be called...

	// We now have to stay alive to keep the magic-link alive...
	for {
		// loop forever...
		time.Sleep(time.Duration(1<<63 - 1))
	}
	return fmt.Errorf("mount: infinite loop exited")
}

func main() {
	app := cli.NewApp()
	app.Name = "memfdexe"
	app.Usage = usage

	// Set version to be the same as runC.
	var v []string
	if version != "" {
		v = append(v, version)
	}
	if gitCommit != "" {
		v = append(v, "commit: "+gitCommit)
	}
	app.Version = strings.Join(v, "\n")

	// Set the flags.
	app.Flags = []cli.Flag{
		cli.BoolFlag{
			Name:  "cleanup",
			Usage: "Do not create a new memfd-sealed file, only clean up an existing one at <path>.",
		},
	}

	app.Action = func(ctx *cli.Context) error {
		args := ctx.Args()
		if len(args) != 1 {
			return errors.New("need to specify a single path to the runc binary")
		}
		path := ctx.Args()[0]

		err := cleanup(path)
		// We only care about cleanup errors when doing --cleanup.
		if ctx.Bool("cleanup") {
			return err
		}
		return mount(path)
	}
	if err := app.Run(os.Args); err != nil {
		fmt.Fprintf(os.Stderr, "memfdexe: %v\n", err)
		os.Exit(1)
	}
}
```

Note that this workaround has two major drawbacks:

1. The program needs to stay alive on the host, meaning it's a daemon you probably need to configure as as a systemd unit.
2. Only programs with the same uid as the daemon (more specifically, programs with ptrace capabilities over the daemon) can access its `/proc/self/fd/...` files, meaning that if you use this workaround then regular users won't be able to run `/usr/bin/runc` directly (this is probably not an issue for Kubernetes or Docker, as I suspect they only ever run `runc` as the same root user).

Reply:
Are there any specific kernel version requirements for using memfdexe? If so, what is the minimum version requirement?

Reply:
@113xiaoji The same as `memfd_create(2)`, which is upstream Linux 3.17 (though I think RHEL backported it to older kernel versions for RHEL 7?).

Note that in order for `memfd-bind` to work (this is what I renamed it in this PR -- see the newset commit) you need to have the second commit (`nsenter: cloned_binary: use MFD_EXEC and F_SEAL_EXEC`) applied to your copy of runc in order for it to correctly handle the seals set by `memfd-bind` on newer kernels.

Reply:
> As for the memory usage, I want to reiterate that the memory usage is _temporary_ -- for most users, `runc init` is only alive for a few hundred milliseconds and then dies soon after. AFAIK `mmap`'d file-backed memory is not accounted by memcg, so there would be no double-accounting -- the total extra memory usage would be 10MB per `runc init` process (for the 100-300ms that the process is alive). Based on my tests, setting the limit to >13MB (the current size of the `runc` binary) is enough for containers to be able to start. However, if you were to do 5000 `runc exec`s concurrently, there could be significant memory usage for the short period of time those processes are running.

@cyphar Hello, I just want to confirm with you that, to be precise, the memory used by memfs is not accounted for by memcg (Memory cgroups). This means that it essentially consumes the host's memory, not the container's memory. Consequently, compared to before the modification, each container operation will only consume additional memory on the host equivalent to the size of the runc binary, which is about 10MB. Am i right?

Reply:
@cyphar In the scenario of a runc upgrade, must the runc replacement be performed first, followed by a restart of memfd-bind?

Reply:
@cyphar 
The primary function of this tool is that when other programs try to open a file via the runc path (e.g., /usr/local/sbin/runc), they are actually opening a symbolic link that points to a memory file descriptor (memfd).

The reason this approach does not consume additional memory is that each time "runc init" is executed, the symbolic link pointing to the memory file descriptor (memfd) is reused, thus eliminating the need for additional memory allocation.

As for the roughly 10MB of memory consumed by this tool, it's because it needs to store the binary file of runc in memory. This memory consumption is attributed to the memory cgroup (memcg) of the host machine, rather than being within the memory limits of the container. Therefore, even in memory-constrained environments, such as inside a container, running runc in this manner will not add extra memory burden.  Am i right? If there are any inaccuracies, please point them out. I greatly appreciate it.

Reply:
After looking into this some more, this binary is not needed -- in cgroupv2, the memfd approach will never affect the container's memory limits (because memory charges are not migrated when process are moved between cgroups in cgroupv2).